### PR TITLE
V0.2.0 internal audit

### PR DESCRIPTION
**Pull request by 77ph**
- Internal audit of latest version
- Please, discussed and addressed

### CODE REVIEW

**Review by kupermind — APPROVED** [review]

File: contracts/MemeFactory.sol [hunk 1]
```diff
@@ -54,8 +55,8 @@ interface IERC20 {
 ///           $SMTH to be burned.
 abstract contract MemeFactory {
     event OLASJourneyToAscendance(uint256 amount);
-    event Summoned(address indexed summoner, address indexed memeToken, uint256 nativeTokenContributed);
-    event Hearted(address indexed hearter, address indexed memeToken, uint256 amount);
+    event Summoned(address indexed summoner, uint256 indexed memeNonce, uint256 nativeTokenContributed);
+    event Hearted(address indexed summoner, uint256 indexed memeNonce, uint256 amount);
```
Comment (DavidMinarsch) [inline contracts/MemeFactory.sol:59]:
```suggestion
    event Hearted(address indexed hearter, uint256 indexed memeNonce, uint256 amount);
```

File: contracts/MemeFactory.sol [hunk 2]
```diff
@@ -201,11 +208,60 @@ abstract contract MemeFactory {
...
+        uint32[] memory secondsAgos = new uint32[](2);
+        // Start of the period
+        secondsAgos[0] = SECONDS_AGO;
+
+        // Fetch the tick cumulative values from the pool
+        (int56[] memory tickCumulatives, ) = IUniswapV3(pool).observe(secondsAgos);
+
+        // Calculate the average tick over the time period
+        int56 tickCumulativeDelta = tickCumulatives[1] - tickCumulatives[0];
+        int24 averageTick = int24(tickCumulativeDelta / int56(int32(SECONDS_AGO)));
```
Comment (DavidMinarsch) [inline contracts/MemeFactory.sol:222]: different ints bottom and top, intended?

File: contracts/MemeFactory.sol [hunk 3]
```diff
@@ -341,24 +392,45 @@ abstract contract MemeFactory {
...
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            memeToken := create2(0x0, add(0x20, payload), mload(payload), randomNonce)
+        }
+
+        // Check for non-zero token address
+        require(memeToken != address(0), "Token creation failed");
+
+        // Update nonce
+        _nonce = localNonce + 1;
```
Comment (DavidMinarsch) [inline contracts/MemeFactory.sol:423]: Why? In summon we increase. Shouldn't increase it here.

Comment (kupermind) [reply contracts/MemeFactory.sol:423]: We need to increase after each operation. Nonce must be different each time.

File: contracts/MemeFactory.sol (hunk 2 above)
Comment (kupermind) [reply contracts/MemeFactory.sol:222]: Why? from line 221 the top is `int56` as well. the result is then brought to `int24`.

Comment (DavidMinarsch) [reply contracts/MemeFactory.sol:222]: ok, sorry didn't see brackets correctly

File: contracts/MemeFactory.sol (hunk 3 above)
Comment (DavidMinarsch) [reply contracts/MemeFactory.sol:423]: No you need to increase nonce once per token, not both at summon and unleash

Comment (kupermind) [reply contracts/MemeFactory.sol:423]: ok, but it does not matter at all

**Pull request merged**